cmd/contour/shutdownmanager.go

```diff
@@ -0,0 +1,132 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/projectcontour/contour/internal/metrics"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/projectcontour/contour/internal/workgroup"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+// handler for /healthz
+func (s *shutdownmanagerContext) healthzHandler(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+	_, err := w.Write([]byte("OK"))
+	if err != nil {
+		s.Error(err)
+	}
+}
+
+// handles /shutdown
+func (s *shutdownmanagerContext) shutdownHandler(w http.ResponseWriter, r *http.Request) {
+	prometheusURL := fmt.Sprintf("http://%s:%d%s", s.envoyHost, s.envoyPort, s.prometheusPath)
+	envoyAdminURL := fmt.Sprintf("http://%s:%d/healthcheck/fail", s.envoyHost, s.envoyPort)
+
+	// Send shutdown signal to Envoy to start draining connections
+	err := shutdownEnvoy(envoyAdminURL)
+	if err != nil {
+		s.Errorf("Error sending envoy healthcheck fail: %v", err)
+	}
+
+	s.Infof("Sent healthcheck fail to Envoy...waiting %s before polling for draining connections", s.checkDelay)
+	time.Sleep(s.checkDelay)
+
+	for {
+		openConnections, err := getOpenConnections(prometheusURL, s.prometheusStat, s.prometheusValues)
+		if err != nil {
+			s.Error(err)
+		} else {
+			if openConnections <= s.minOpenConnections {
+				s.Infof("Found %d open connections with min number of %d connections. Shutting down...", openConnections, s.minOpenConnections)
+				return
+			}
+			s.Infof("Found %d open connections with min number of %d connections. Polling again...", openConnections, s.minOpenConnections)
+		}
+		time.Sleep(s.checkInterval)
+	}
+}
+
+// shutdownEnvoy sends a POST request to /healthcheck/fail to tell Envoy to start draining connections
+func shutdownEnvoy(url string) error {
+	resp, err := http.Post(url, "", nil)
+	if err != nil {
+		return fmt.Errorf("creating POST request for URL %q failed: %s", url, err)
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("POST request for URL %q returned HTTP status %s", url, resp.Status)
+	}
+	return nil
+}
+
+// getOpenConnections parses a http request to a prometheus endpoint returning the sum of values found
+func getOpenConnections(url, prometheusStat string, prometheusValues []string) (int, error) {
+	// Make request to Envoy Prometheus endpoint
+	resp, err := http.Get(url)
+	if err != nil {
+		return -1, fmt.Errorf("GET request for URL %q failed: %s", url, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return -1, fmt.Errorf("GET request for URL %q returned HTTP status %s", url, resp.Status)
+	}
+
```

Comment:
Nit: getOpenConnections and getOpenConnections have identical structure but are formatted differently, do you mind aligning them?